azure/services/scalesets/scalesets.go

```diff
@@ -85,32 +83,68 @@ func (s *Service) Reconcile(ctx context.Context) (retErr error) {
 	ctx, log, done := tele.StartSpanWithLogger(ctx, "scalesets.Service.Reconcile")
 	defer done()
 
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
+
 	if err := s.validateSpec(ctx); err != nil {
 		// do as much early validation as possible to limit calls to Azure
 		return err
 	}
 
 	var err error
+	var resultErr error
 
-	scaleSetSpec := s.Scope.ScaleSetSpec()
+	spec := s.Scope.ScaleSetSpec(ctx)
 
-	// check if there is an ongoing long running operation
-	var fetchedVMSS *azure.VMSS
-	future := s.Scope.GetLongRunningOperationState(s.Scope.ScaleSetSpec().Name, serviceName, infrav1.PutFuture)
-	if future == nil {
-		future = s.Scope.GetLongRunningOperationState(s.Scope.ScaleSetSpec().Name, serviceName, infrav1.PatchFuture)
+	vmssInstances, err := s.Client.ListInstances(ctx, spec.ResourceGroupName(), spec.ResourceName())
+	if err != nil {
+		resultErr = errors.Wrapf(err, "failed to get existing instances")
+		s.Scope.UpdatePutStatus(infrav1.BootstrapSucceededCondition, serviceName, resultErr)
+		return resultErr
 	}
 
-	defer func() {
-		// save the updated state of the VMSS for the MachinePoolScope to use for updating K8s state
-		if fetchedVMSS == nil {
-			fetchedVMSS, err = s.getVirtualMachineScaleSet(ctx, scaleSetSpec.Name)
-			if err != nil && !azure.ResourceNotFound(err) {
-				log.Error(err, "failed to get vmss in deferred update")
-			}
+	scaleSetSpec, ok := spec.(*ScaleSetSpec)
+	if !ok {
+		return errors.Errorf("%T is not of type ScaleSetSpec", spec)
+	}
+	scaleSetSpec.VMSSInstances = vmssInstances
+	scaleSetSpec.VMImage, err = s.Scope.GetVMImage(ctx)
+	if err != nil {
+		resultErr = errors.Wrapf(err, "failed to get vm image")
+		s.Scope.UpdatePutStatus(infrav1.BootstrapSucceededCondition, serviceName, resultErr)
+		return resultErr
+	}
+	s.Scope.SaveVMImageToStatus(scaleSetSpec.VMImage)
+
+	scaleSetSpec.BootstrapData, err = s.Scope.GetBootstrapData(ctx)
+	if err != nil {
+		resultErr = errors.Wrapf(err, "failed to get bootstrap data")
+		s.Scope.UpdatePutStatus(infrav1.BootstrapSucceededCondition, serviceName, resultErr)
+		return resultErr
+	}
+	scaleSetSpec.MaxSurge, err = s.Scope.MaxSurge()
+	if err != nil {
+		resultErr = errors.Wrapf(err, "failed to get max surge")
+		s.Scope.UpdatePutStatus(infrav1.BootstrapSucceededCondition, serviceName, resultErr)
+		return resultErr
+	}
+
+	result, err := s.CreateOrUpdateResource(ctx, scaleSetSpec, serviceName)
+	s.Scope.UpdatePutStatus(infrav1.BootstrapSucceededCondition, serviceName, err)
+
+	if err == nil && result != nil {
+		vmss, ok := result.(compute.VirtualMachineScaleSet)
+		if !ok {
+			return errors.Errorf("%T is not a compute.VirtualMachineScaleSet", result)
 		}
 
+		fetchedVMSS := converters.SDKToVMSS(vmss, vmssInstances)
+		// Note: converters.SDKToVMSS() will never return nil but we should check anyway
```

Comment:
maybe that func shouldn't be returning a pointer if it never returns nil?

Reply:
I think that make sense. Do you think we should do that in this PR or a separate follow up?